### PR TITLE
[Merged by Bors] - asset cards with more metadata

### PR DESCRIPTION
**Pull request by IceSentry**
## Objective

Update the asset cards to support more metadata. The main goal is to be able to show the supported bevy version. Everything else is secondary right now.

## Solution

Support more metadata by using the styling from https://github.com/bevyengine/bevy-website/issues/231.

I'm missing a few things, like different colors for bevy versions, but that's because I don't know how to do that with zola templates. I also ignored the author name and crate version for now because it doesn't apply to every crate and it can be iterated on in the future.

image

Supported metadata:
- `bevy_versions`: array of supported version
- `licenses`: array of licenses

The metadata currently doesn't exist, but the code is written in such a way to not break if it doesn't exist. I created #397 to generate the required metadata.

A lot of the values were just found by me playing around with values and finding something that looked good. I'm absolutely opened to changing them. To be clear, a lot of what I have right now are just values that ended up working, but there's probably a better way to do most of those things.

## Remaining questions

- ~~What to do with assets without images.~~
  - Solved: I added a gradient to make it a bit less empty.
- How to handle different color based on the value of the version tag by using zola templates

## Notes

Until we add the required metadata, the cards will have a lot of empty space at the bottom, I already have a prototype ready to add the missing metadata, so that should be fixed pretty fast.

### CODE REVIEW

**Comment by IceSentry** [comment]
@TimJentzsch since I can't figure out how to request a review, hopefully this ping will do the job.

File: sass/components/_card.scss [hunk 1]
```diff
@@ -85,4 +85,4 @@
     margin-left: auto;
     margin-right: auto;
     align-items: center;
-}
+}
```
Comment (mockersf) [inline sass/components/_card.scss:88]:
```suggestion
}

```

**Comment by mockersf** [comment]
How it looks for me (Safari and Firefox):
Screenshot 2022-07-18 at 12 06 49

* The icons seems to not be aligned with the centre of the tags
* The license icon is too close to the license tag
* If there are no version, the license is on the left

There is also an issue when there are too many bevy versions, this would hide the licenses. That could be solved later as it's unlikely to be an issue today

For the remaining questions, I would like this to move without solving them:
* Assets without an image: let's keep them empty for now, and take the time to find a good placeholder
* Colors for versions: colors are... too complicated. This info is needed now, even without colors

**Comment by TimJentzsch** [comment]
> - The icons seems to not be aligned with the centre of the tags
> - The license icon is too close to the license tag
> - If there are no version, the license is on the left

We should be able to fix all of that with :sparkles: flexbox :sparkles: 

The main ingredients:

- `display: flex`: This is where the fun begins!
- `justify-content: space-between`: The first item goes to the start, the second item goes to the end.
- Wrap the version and license each in a `div`. Only the content of the `div` is conditional, that way you keep the license to the right.
- `align-items: center`: Aligns the icons at the center of the tags.
- `gap: 5px`: Space between icon and tag name.
- `flex-flow: row-wrap`: If there are too many bevy versions, the license wraps in the next row.

That should be everything we need. In a couple of hours I might also have time to do a mockup with vanilla HTML and CSS if needed.

**Comment by afonsolage** [comment]
Since the new layout always expects an image (like a thumbnail), maybe add a [default bevy image](https://bevyengine.org/processed_images/4e424e8bfdfe1d4800.png) when there is no image on metadata?

image

The current layout adapts when there is no image.

**Comment by mockersf** [comment]
> Since the new layout always expects an image (like a thumbnail), maybe add a [default bevy image](https://bevyengine.org/processed_images/4e424e8bfdfe1d4800.png) when there is no image on metadata?

Yes, we'll need a placeholder, but not the Bevy logo

**Comment by TimJentzsch** [comment]
> The current layout adapts when there is no image.

In my opinion that makes more sense than having a big gray box that is not used

**Comment by IceSentry** [comment]
For the image thing, I was mostly questioning what to do right now, but yes in the future we should either make images mandatory or at the very least some sort of generated placeholder. I believe once cards are more clearly used to show case images we'll see more people submitting images. Could be worth explicitly calling out in the blog post.

**Comment by IceSentry** [comment]
For the bevy and license logo, that was me playing around with it and breaking it after making my original PR it was super late though so I stopped and went to bed. I'll try to fix it later today.

**Comment by doup** [comment]
It's looking good!

For the image adding some light background image (maybe some grey cogs?) + the name of the asset on top should look good enough. I think that removing it, with this vertical layout, is going to look ugly as the title/text will be misaligned between cards.

I have a couple of minor nits on margins/padding/alignment. On top of what `@mockersf` already mentioned in https://github.com/bevyengine/bevy-website/pull/394#issuecomment-1187016606, I would say that the margins/padding would look better if they were more consistent:
- Reduce bottom gap of the card (on the current online version on both assets/examples cards the same gap is used for both axis). You probably have some `margin-bottom`; just removing it should work as the parent `grid` container is responsible for the gap.
- Increase left padding on asset title to match description
- Increase bottom padding for de card

image

Also, on your screenshot it looks like the text is not vertically centered on the "pills":

image

**Comment by IceSentry** [comment]
Alright, I rewrote most of the styling, but this time I used css grid and it made everything way easier. I tried to apply most of the feedback that I saw here.

I also added a gradient for the assets without images just so it's less empty

image

@doup sorry I didn't see your feedback before committing, but I think I fixed all of that

File: sass/pages/_assets.scss [hunk 2]
```diff
@@ -90,22 +90,34 @@
...
                 background-color: rgb(110, 110, 247);
             }
         }
-
-        a {
-            text-decoration: none;
-            color: $link-color;
-        }
     }
 }
```
Comment (TimJentzsch) [inline sass/pages/_assets.scss]: Nit: Can we add a trailing linebreak here? GitHub complains about that

File: templates/macros/assets.html [hunk 3]
```diff
@@ -1,56 +1,94 @@
...
-<div class="card" id="{{ post.title | slugify }}">
-    <a href="{{ post.extra.link }}">
-        {% if post.extra.image %}
-        <div class="card-image card-image-default">
+<div id="{{ post.title | slugify }}">
+    <a class="asset-card" href="{{ post.extra.link }}">
+        <div class="asset-card__banner">
+            <div class="asset-card__title">{{post.title}}</div>
+            {% if post.extra.image %}
             <img src="{{ image_macros::resize_image(path=post.extra.image, width=340, height=340) }}" />
```
Comment (TimJentzsch) [inline templates/macros/assets.html:31]: This seems to be missing an `alt` attribute`

File: templates/macros/assets.html [hunk 4]
```diff
@@ -1,56 +1,94 @@
...
+                {% for version in post.extra.bevy_versions %}
+                <span class="asset-card__tag">{{version}}</span>
+                {% endfor %}
+            </div>
+            {% endif %}
+            {% if post.extra.licenses %}
+            <div class="asset-card__tag-list asset-card__licenses">
+                {{ assets_macros::license_logo_svg() }}
+                {% for license in post.extra.licenses %}
+                <span class="asset-card__tag">{{license}}</span>
```
Comment (TimJentzsch) [inline templates/macros/assets.html:50]: Similar here, except that `alt` attributes don't work on `svg`s.
From what I've read, the best way to do this is to add a `title` element inside the `svg`: [Blog post](https://css-tricks.com/accessible-svgs/#aa-2-inline-svg), [W3 Reference](https://www.w3.org/TR/SVG11/struct.html#DescriptionAndTitleElements)

File: templates/macros/assets.html [hunk 5]
```diff
@@ -1,56 +1,94 @@
...
+</li>
 {% endmacro %}
+
+{% macro assets_menu(root, prefix) %}
+<ul class="tree-menu">
+    {% for s in root.subsections %}
+    {{ self::assets_menu_row(prefix=prefix, section_path=s)}}
+    {% endfor %}
+</ul>
+{% endmacro %}
```
Comment (TimJentzsch) [inline templates/macros/assets.html:94]: Nit: Trailing line break is missing here

File: sass/pages/_assets.scss (hunk 2 above)
Comment (IceSentry) [reply sass/pages/_assets.scss]: Is it possible you didn't refresh? My latest commit should have already fixed that one

File: templates/macros/assets.html [hunk 6]
```diff
@@ -1,56 +1,94 @@
...
+    <div class="{{label_class}}">
+        <a class="tree-menu__link" href="#{{section.title | slugify}}">{{ section.title }}</a>
         {% if section.subsections %}
-            <ul class="tree-menu">
-                {% for s in section.subsections %}
-                    {{ self::assets_menu_row(prefix=prefix, section_path=s) }}
-                {% endfor %}
-            </ul>
+        <label class="tree-menu__toggle" for="{{id}}">
+            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg">
```
Comment (TimJentzsch) [inline templates/macros/assets.html:74]: This is missing an `alt` attribute and should maybe have a `title` attribute to indicate that you can toggle something here (if I understand this correctly)

**Review by TimJentzsch — CHANGES_REQUESTED** [review]
The screenshots you sent look good to me!

I did a review run on accessibility issues and found a few, would be nice to fix those before merging :)

File: templates/macros/assets.html (hunk 4 above)
Comment (IceSentry) [reply templates/macros/assets.html:50]: I want to convert the svg to an image asset like the bevy logo anyway so that will work

File: templates/macros/assets.html [hunk 7]
```diff
@@ -1,56 +1,94 @@
...
-                {{ post.description | striptags | truncate(length=100) | safe }}
+        <div class="asset-card__description">
+            {{ post.description | striptags | truncate(length=120) | safe }}
+        </div>
+        <div class="asset-card__tags">
+            {% if post.extra.bevy_versions %}
+            <div class="asset-card__tag-list asset-card__bevy-versions">
+                <img src="/assets/bevy_icon_dark.svg" alt="Small bevy icon">
+                {% for version in post.extra.bevy_versions %}
+                <span class="asset-card__tag">{{version}}</span>
```
[the remark this answers is not in the document]
Comment (IceSentry) [reply templates/macros/assets.html:42]: I literally had that exact "Supported bevy versions" alt text earlier, I have no idea why I changed it lol and yeah, using a title is a good call.

File: templates/macros/assets.html [hunk 8]
```diff
@@ -37,7 +24,7 @@
         <div class="asset-card__tags">
             {% if post.extra.bevy_versions %}
             <div class="asset-card__tag-list asset-card__bevy-versions">
-                <img src="/assets/bevy_icon_dark.svg" alt="Small bevy icon">
+                <img src="/assets/bevy_icon_dark.svg" title="Supported bevy versions">
```
Comment (TimJentzsch) [inline templates/macros/assets.html:27]: Nit: we need both the `alt` and `title`, as the `alt` is also rendered when the `img` can't be loaded.

(Sorry, I should have made that clearer in my comment)

File: templates/macros/assets.html [hunk 9]
```diff
@@ -71,7 +58,8 @@
         <a class="tree-menu__link" href="#{{section.title | slugify}}">{{ section.title }}</a>
         {% if section.subsections %}
         <label class="tree-menu__toggle" for="{{id}}">
-            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg">
+            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg"
+                title="Toggle {{section.title}} section">
```
Comment (TimJentzsch) [inline templates/macros/assets.html:62]: Same here, we also need the `alt` attribute.

File: templates/macros/assets.html [hunk 10]
```diff
@@ -21,14 +8,14 @@
 </svg>
 {% endmacro license_logo_svg %}
 
-
 {% macro card(post) %}
 <div id="{{ post.title | slugify }}">
     <a class="asset-card" href="{{ post.extra.link }}">
```
Comment (TimJentzsch) [inline templates/macros/assets.html:13]: The [MDN docs for `a`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) state:

> Content within each `<a>` **should** indicate the link's destination.

Such that users with visual impairments/using screen-readers can tell where they are going.
I assume that this is not the case here, because we are wrapping a `div` element.

However, I could not find any guidelines to make this accessible. If anyone knows anything, please let me know!

Maybe it's fine though because of the card title that's deeper in the hierarchy?

**Review by TimJentzsch — CHANGES_REQUESTED** [review]

File: templates/macros/assets.html (hunk 10 above)
Comment (TimJentzsch) [reply templates/macros/assets.html:13]: I think they mean that the content of the `<a>` should indicate where the link leads to: [Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#strong_link_text)

**Review by TimJentzsch — APPROVED** [review]
Awesome, looks good to me!

File: sass/pages/_assets.scss [hunk 11]
```diff
@@ -18,29 +18,109 @@
...
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &__description {
+            grid-area: desc;
+            display: flex;
+
+            padding: 0.3rem;
+            font-size: 1rem;
+        }
```
Comment (doup) [inline sass/pages/_assets.scss:77]: I've tried with a little bit of extra padding on the title/description (0.6rem ~ 8px), what do you think?

```suggestion
        &__title {
            position: absolute;
            width: 100%;
            bottom: 0;

            padding: 0.6rem;
            font-weight: bold;
            font-size: 1.2rem;
            background-color: rgba(0, 0, 0, 0.5);
        }

        &__description {
            grid-area: desc;
            display: flex;

            padding: 0.6rem;
            font-size: 1rem;
        }
```

image

File: sass/pages/_assets.scss [hunk 12]
```diff
@@ -18,29 +18,109 @@
         margin-bottom: 52px;
     }
 
-    .card {
-        margin: 0;
-        height: 8rem;
-    }
+    .asset-card {
```
Comment (doup) [inline sass/pages/_assets.scss:21]: I would move `asset-card` into it's own file `sass/components/_asset-card.scss`.
Pages folder (`pages/*`) is kind-of legacy; where both BEM and non-BEM gets mixed.

Comment (IceSentry) [reply sass/pages/_assets.scss:21]: Sure, I'm still new to BEM so wasn't sure how file structure should work so I just updated the existing file. I'll do that tomorrow.

File: sass/pages/_assets.scss [hunk 13]
```diff
@@ -18,29 +18,109 @@
...
+                background-color: $default-image-background-color;
+            }
+        }
+
+        &__title {
+            position: absolute;
+            width: 100%;
+            bottom: 0;
+
+            padding: 0.6rem;
```
Comment (doup) [inline sass/pages/_assets.scss:65]: This shrinks the vertical padding for the title, this way it doesn't look so "airy"… and I stop. 😇 
Anyway, if you think `0.6rem` is too much please reduce as you wish… it's just that I find `0.3rem` too tight.

```suggestion
            padding: 0.3rem 0.6rem;
```

image

File: sass/pages/_assets.scss (hunk 11 above)
Comment (doup) [reply sass/pages/_assets.scss]: Sure, go for it. `0.6` was a suggestion.

**Review by doup — APPROVED** [review]

File: sass/pages/_assets.scss [hunk 14]
```diff
@@ -18,29 +18,109 @@
...
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &__description {
+            grid-area: desc;
+            display: flex;
+
+            padding: 0.6rem;
+            font-size: 1rem;
+        }
```
Comment (doup) [inline sass/pages/_assets.scss:77]: I've noticed that this doesn't handle extra long text correctly.

**Option A**: use 4 lines ellipsis ([the mixin uses this snippet](https://css-tricks.com/almanac/properties/l/line-clamp/)).

```suggestion
        &__description {
            @include text-ellipsis(4);

            grid-area: desc;

            padding: 0.6rem;
            font-size: 1rem;
        }
```

**Option B**, see comment:

File: sass/pages/_assets.scss [hunk 15]
```diff
@@ -18,29 +18,109 @@
...
-        margin: 0;
-        height: 8rem;
-    }
+    .asset-card {
+        display: grid;
+        grid-template-areas:
+            "banner"
+            "desc"
+            "tags";
+        grid-template-rows: 12rem 5.5rem 2rem;
```
Comment (doup) [inline sass/pages/_assets.scss:27]: **Option B**, for comment: https://github.com/bevyengine/bevy-website/pull/394/files#r924157060

```suggestion
        grid-template-rows: 12rem minmax(5.5rem, 1fr) 2rem;
        height: 100%;
```

- `minmax` to allow the description grow, but keep a minimum height
- `height` for the other cards, so that they grow to match the one with the long description

image

File: sass/pages/_assets.scss (hunk 13 above)
Comment (IceSentry) [reply sass/pages/_assets.scss:65]: 😍 I love it, I that was the missing part.

File: sass/pages/_assets.scss (hunk 15 above)
Comment (ickk) [reply sass/pages/_assets.scss]: I'm biased against growing the cards vertically; when we were designing the initial mockups we had a lot of discussion about the amount of text that would be appropriate for cards, and we landed on the ratios and everything that we did based on a relatively little amount of text. 

Descriptions should be short and to the point. Longer exposition about a particular asset belongs on it's asset page (whenever we build that feature) or on the crates.io/repo/webpage linked to imo

Comment (doup) [reply sass/pages/_assets.scss]: This was just an option I was giving him, related comment: https://github.com/bevyengine/bevy-website/pull/394/files#r924157060
He used **"Option A"** in the end, which is just "ellipsis at the fourth line": https://github.com/bevyengine/bevy-website/pull/394/commits/1a93da13e2cdbe0d8d8b2ac4949dba917e76f775

File: sass/pages/_assets.scss (hunk 14 above)
Comment (doup) [reply sass/pages/_assets.scss]: Sorry, I didn't realize, I saw it a after you removed it again on the comment. 🤦

**Comment by alice-i-cecile** [comment]
@IceSentry this needs merge conflict resolution, but IMO we should ship this and iterate.

**Comment by IceSentry** [comment]
I'd rather shipping the metadata loading first. That way the cards will have actual data to show. Right they will just look weirdly empty

**Comment by alice-i-cecile** [comment]
Sounds good to me. I'll mark this as Blocked so we don't forget.

**Comment by IceSentry** [comment]
Well, technically it isn't blocked, it could easily be merged without any issue, it would just look a bit empty at the bottom since there would be no license or bevy versions

**Comment by alice-i-cecile** [comment]
Technically, but I'd be against it :p

**Comment by IceSentry** [comment]
The other PR is ready it just needs more reviews. it's just that most people don't look at `bevy-website` 😦

File: sass/components/_asset-card.scss [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+        background-color: $card-hover-background;
+        border: 2px solid #6b6b6b;
+    }
+
+    &__banner {
+        grid-area: banner;
+        display: block;
+        position: relative;
+
+        background: linear-gradient(45deg,
```
Comment (cart) [inline sass/components/_asset-card.scss:26]: Totally a matter of opinion, but I'd like to propose a more subtle / aligned gradient: `linear-gradient(0deg, #37393e 0%, #1b1b1b 100%)`

image

The current one draws the eye a bit too much and feels a bit aggressive / like a bright light is shining on something very metallic / reflective. And notably with that analogy the light should behave differently for each tile based on its position relative to your eye, but it doesn't. Which also contributes to my (personal) impression that it feels "off":

image

File: sass/components/_asset-card.scss [hunk 17]
```diff
@@ -0,0 +1,107 @@
+.asset-card {
+    display: grid;
+    grid-template-areas:
+        "banner"
+        "desc"
+        "tags";
+    grid-template-rows: 12rem 5.5rem 2rem;
```
Comment (cart) [inline sass/components/_asset-card.scss:7]: I think we should probably rein in the height of the text section a bit (cut down on the character limit if necessary, maybe reduce the font size). The cards have _just_ a bit too much height right now. Most descriptions are short and this results in _a lot_ of wasted space. I'm down to do that in a separate PR though.

**Review by cart — COMMENTED** [review]
I'm loving these changes!

File: sass/components/_asset-card.scss (hunk 16 above)
Comment (IceSentry) [reply sass/components/_asset-card.scss:26]: Good idea, I like it, a lot more subtle while still being nicer than a solid color.

File: sass/components/_asset-card.scss (hunk 17 above)
Comment (IceSentry) [reply sass/components/_asset-card.scss:7]: I'll see what I can do. I picked the height based on the longer descriptions but it was mostly just playing with some values. I'm down with reducing the size a bit

Comment (IceSentry) [reply sass/components/_asset-card.scss:7]: I reduced the padding a bit and limited it to 3 lines now. This is how it looks.
image

I tried going to 2 lines max but it looked weird.

**Comment by rparrett** [comment]
I think that we may have lost lazy-loading of the asset images in the rebase.

File: templates/macros/assets.html [hunk 18]
```diff
@@ -1,56 +1,93 @@
...
-    <a href="{{ post.extra.link }}">
-        {% if post.extra.image %}
-        <div class="card-image card-image-default">
-            <img src="{{ image_macros::resize_image(path=post.extra.image, width=340, height=340) }}" loading="lazy" />
+<div id="{{ post.title | slugify }}">
+    <a class="asset-card" href="{{ post.extra.link }}">
+        <div class="asset-card__banner">
+            <div class="asset-card__title">{{post.title}}</div>
+            {% if post.extra.image %}
+            <img src="{{ image_macros::resize_image(path=post.extra.image, width=340, height=340) }}"
```
Comment (rparrett) [inline templates/macros/assets.html:21]:
```suggestion
            <img src="{{ image_macros::resize_image(path=post.extra.image, width=370, height=370) }}"
```

File: sass/components/_asset-card.scss (hunk 17 above)
Comment (cart) [reply sass/components/_asset-card.scss]: I dig it! One more small tweak: I do think adding back a bit more padding is "worth it".

File: templates/macros/assets.html [hunk 19]
```diff
@@ -1,11 +1,16 @@
...
-    <path d="M554.9 154.5c-17.62-35.25-68.12-35.38-85.87 0c-87 174.3-84.1 165.9-84.1 181.5c0 44.13 57.25 80 128 80s127.1-35.88 127.1-80C639.1 319.9 641.4 327.3 554.9 154.5zM439.1 320l71.96-144l72.17 144H439.1zM256 336c0-16.12 1.375-8.75-85.12-181.5c-17.62-35.25-68.12-35.38-85.87 0c-87 174.3-84.1 165.9-84.1 181.5c0 44.13 57.25 80 127.1 80S256 380.1 256 336zM127.9 176L200.1 320H55.96L127.9 176zM495.1 448h-143.1V153.3C375.5 143 393.1 121.8 398.4 96h113.6c17.67 0 31.1-14.33 31.1-32s-14.33-32-31.1-32h-128.4c-14.62-19.38-37.5-32-63.62-32S270.1 12.62 256.4 32H128C110.3 32 96 46.33 96 64S110.3 96 127.1 96h113.6c5.25 25.75 22.87 47 46.37 57.25V448H144c-26.51 0-48.01 21.49-48.01 48c0 8.836 7.165 16 16 16h416c8.836 0 16-7.164 16-16C544 469.5 522.5 448 495.1 448z"/>
+{% macro init_svg() %}
+<svg style="display: none" version="2.0">
+    <defs>
+        <symbol id="asset-card__scale-balanced-svg" viewBox="0 0 640 512">
+            <!--! Font Awesome Pro 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2022 Fonticons, Inc. -->
```
Comment (cart) [inline templates/macros/assets.html:7]: Hmm this is explicitly calling itself out as "font awesome pro". We shouldn't be using any icons from there (and I dont think this one is part of pro).

Comment (cart) [reply templates/macros/assets.html:7]: https://fontawesome.com/license/free

Comment (IceSentry) [reply templates/macros/assets.html:7]: That's the exact svg with the same comment I get when I copy it from there https://fontawesome.com/icons/scale-balanced?s=solid

I'm not sure why it says it's under the pro license

File: sass/components/_asset-card.scss (hunk 17 above)
Comment (IceSentry) [reply sass/components/_asset-card.scss]: I increased the padding a bit and made the area for the tags slightly larger so it could have a more consistent padding.

image

Comment (cart) [reply sass/components/_asset-card.scss]: Lovely!

File: templates/macros/assets.html (hunk 19 above)
Comment (cart) [reply templates/macros/assets.html]: If you download the "free" package, the svgs have the correct license embedded in them. I just pushed that svg content.

Comment (IceSentry) [reply templates/macros/assets.html]: Interesting, it's a bit weird that they have the wrong license on the site for the exact same svg. That's good to know though so I downloaded it for the future.
